Detectors/MUON/MCH/DevIO/Digits/digits-random-generator-workflow.cxx

```diff
@@ -0,0 +1,212 @@
+    for (auto deid : deids) {
+      o2::mch::mapping::Segmentation seg(deid);
+      int16_t nofPads = static_cast<int16_t>(seg.nofPads());
+      npads[i++] = nofPads;
+    }
+    first = false;
+  }
+  return npads;
+}
+} // namespace
+
+constexpr const char* OPTNAME_NOF_ROFS_PER_TF = "nof-rofs-per-tf";
+constexpr const char* OPTNAME_OCCUPANCY = "occupancy";
+constexpr const char* OPTNAME_SEED = "seed";
+
+class DigitGenerator : public o2::mch::io::DigitIOBaseTask
+{
+ public:
+  void init(InitContext& ic)
+  {
+    io::DigitIOBaseTask::init(ic); // init common options
+    mOccupancy = ic.options().get<float>(OPTNAME_OCCUPANCY);
+    if (mOccupancy <= 0.0 || mOccupancy > 1.0) {
+      throw std::invalid_argument("occupancy must be between >0 and <=1");
+    }
+    mNofRofPerTimeFrame = ic.options().get<int>(OPTNAME_NOF_ROFS_PER_TF);
+    mSeed = ic.options().get<int>(OPTNAME_SEED);
+
+    if (!mSeed) {
+      std::random_device rd;
+      mSeed = rd();
+    }
+    mMersenneTwister.seed(mSeed);
+
+    LOGP(INFO,
+         "Will generate {:7.2f}% of pads in {:4d} ROFs "
+         "per timeframe, for {:4d} timeframes",
+         mOccupancy * 100.0, mNofRofPerTimeFrame, mMaxNofTimeFrames);
+  }
+
+  void run(ProcessingContext& pc)
+  {
+    if (mNofProcessedTFs >= mMaxNofTimeFrames) {
+      pc.services().get<ControlService>().endOfStream();
+      return;
+    }
+
+    if (shouldProcess()) {
+      incNofProcessedTFs();
+      std::vector<Digit> digits;
+      std::vector<ROFRecord> rofs;
+      int ndigits{0};
+
+      for (auto i = 0; i < mNofRofPerTimeFrame; i++) {
+        auto n = generateRandomDigits(mOccupancy, digits);
+        o2::InteractionRecord ir{mBC, mOrbit++};
+        rofs.emplace_back(ir, ndigits, n);
+        ndigits += n;
+        mOrbit++;
+      }
+      pc.outputs().snapshot(OutputRef{"rofs"}, rofs);
+      pc.outputs().snapshot(OutputRef{"digits"}, digits);
+      printSummary(digits, rofs);
+      printFull(digits, rofs);
+    }
+    incTFid();
+  }
+
+  /**
+ * Populate the digits vector with (Nmch * occupancy) digits
+ * where Nmch is the total number of channels in MCH (1064008)
+ * 
+ * Each member variable of the digit is uniformly distributed within
+ * its expected range, as an attempt to maximize the entropy of the
+ * generated digits, which can be handy to test the efficiency of the
+ * entropy encoder used to create the Compressed Time Frame (CTF) for instance.
+ *
+ * @param occupancy is a number between 0 and 1.
+ * @digits a vector where the generated digits will be appended. 
+ * That vector is not cleared by this function, so digits can 
+ * be accumulated if need be.
+ * @returns the number of digits added to the input digits vector.
+ */
+  int generateRandomDigits(float occupancy,
+                           std::vector<Digit>& digits)
+  {
+    int n{0};
+    std::uniform_int_distribution<int32_t> adc{0, 1024 * 1024};
+    std::uniform_int_distribution<int32_t> tfTime{0, 512 * 3564};
```

Comment:
I think tfTime should be always set to 0 since we decided to drop this info in the digits (at least at the decoder level where the digit time should be given by the ROF IR).

In any case, a tfTime randomly generated in the full range of the TF will not be consistent with the ROF IR.

Reply:
Agreed, but until we actually remove the time from the digit, I'd prefer to keep it there as well.